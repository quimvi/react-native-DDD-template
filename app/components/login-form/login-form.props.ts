import { ViewStyle } from 'react-native';

export interface LoginFormProps {
  /**
   * True if and error during login event is present.
   */
  title: string;

  /**
   * Callback for login action
   */
  onLogin?: () => void;

  /**
   * True if and error during login event is present.
   */
  error?: string;

  /**
   * An optional style override useful for padding & margin.
   */
  style?: ViewStyle;
}
