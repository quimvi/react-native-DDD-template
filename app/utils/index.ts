/**
 * Parse a string to boolean.
 *
 * @param {string} bool String to check.
 */
export function parseBoolean(bool: string) {
  return ['1', 'True', 'true'].includes(bool);
}
