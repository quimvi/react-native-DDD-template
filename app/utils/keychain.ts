import * as ReactNativeKeychain from 'react-native-keychain';

/**
 * Saves some credentials securely.
 *
 * @param username The username
 * @param password The password
 * @param server The server these creds are for.
 */
export async function save(username: string, password: string, server?: string) {
  if (server) {
    await ReactNativeKeychain.setInternetCredentials(server, username, password);
    return true;
  } else {
    return ReactNativeKeychain.setGenericPassword(username, password);
  }
}

/**
 * Loads credentials that were already saved.
 *
 * @param server The server that these creds are for
 */
export async function load(server?: string) {
  if (server) {
    const creds = await ReactNativeKeychain.getInternetCredentials(server);
    return {
      password: creds ? creds.password : null,
      server,
      username: creds ? creds.username : null,
    };
  } else {
    const creds = await ReactNativeKeychain.getGenericPassword();
    if (typeof creds === 'object') {
      return {
        password: creds.password,
        server: null,
        username: creds.username,
      };
    } else {
      return {
        password: null,
        server: null,
        username: null,
      };
    }
  }
}

/**
 * Resets any existing credentials for the given server.
 *
 * @param server The server which has these creds
 */
export async function reset(server?: string) {
  if (server) {
    await ReactNativeKeychain.resetInternetCredentials(server);
    return true;
  } else {
    const result = await ReactNativeKeychain.resetGenericPassword();
    return result;
  }
}
