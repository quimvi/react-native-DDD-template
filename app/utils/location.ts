export function toRadians(value: number): number {
  return (value * Math.PI) / 180;
}

export const EARTH_RADIUS_METERS = 6371000;
