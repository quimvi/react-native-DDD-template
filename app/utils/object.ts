import { mergeAll, flatten, reduce, assoc, keys } from 'ramda';

/**
 * Search in an object the key corresponding to a value
 *
 * @param obj Object where we are going to search the key
 * @param value Value to search
 */
export function getKeyByValue(obj: object, value: any) {
  return Object.keys(obj).find((key) => obj[key] === value);
}

/**
 * Creates a list of string from an object, to represent all the keys
 * with the format ['key1.key2','key1.key3']
 *
 * @param obj Object to generate the list
 * @param keyPrefix Add a prefix to each key
 * @param separator Separator to use between keys
 */
export function extractKeys(obj: object, keyPrefix = '', separator = '.'): string[] {
  const combinedKeys = [];
  const keys = Object.keys(obj);

  keys.forEach((key) => {
    if (typeof obj[key] === 'string') {
      if (key.includes('_plural')) {
        return;
      }
      combinedKeys.push(keyPrefix + key);
    } else {
      combinedKeys.push(...extractKeys(obj[key], keyPrefix + key + separator));
    }
  });

  return combinedKeys;
}

/**
 * Creates a new object formed by the keys which has the attribute passed.
 * Each key value will be the value of that attribute.
 *
 * @param obj Target object which has the desired attribute
 * @param attribute Attribute to return.
 */
export function reduceKeyByAttribute(obj: object, attribute: string) {
  return Object.keys(obj).reduce(function (accum, key) {
    const item = obj[key];

    if (attribute in item) {
      accum[key] = item[attribute];
    }

    return accum;
  }, {});
}

/**
 * Mix two or more objects.
 *
 * @param {object[]} objects The array of objects to mix.
 */
export const enhance = <T extends object>(objects: object[]): T => {
  return mergeAll(flatten(objects)) as T;
};

/**
 * If element is an array, return the element[index]. If not, return element
 * @param element Array or default value for the preset.
 * @param index Item to access in case preset is an array.
 */
export function getElement<T>(element: T | T[], index: number): T {
  if (Array.isArray(element)) {
    return element.length > index ? element[index] : null;
  } else {
    return element;
  }
}

/**
 * Given an array an a map of key:string -> value:string, rename the key properties in each
 * object in the array for the value.
 * @param array Array to rename
 * @param keysMap Map of renamings
 */
export function renameKey<T>(array: T[], keysMap: Record<string, string>): T[] {
  return array.map((obj: T) =>
    reduce((acc, key) => assoc(keysMap[key] || key, obj[key], acc), {} as T, keys(obj)),
  );
}
