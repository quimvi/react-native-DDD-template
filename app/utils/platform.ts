import { Platform } from 'react-native';

/**
 * Returns the Android value if the platform is Android, and the ios value if the platform is iOS
 * @param android Value to return if is Android
 * @param platform Value to return if is iOS
 */
function select<T>(android: T, ios: T): T {
  return Platform.select({
    android,
    ios,
  });
}

export default { select };
