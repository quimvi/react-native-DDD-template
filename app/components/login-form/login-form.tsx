import * as React from 'react';
import { Keyboard, TouchableWithoutFeedback, View, Button } from 'react-native';

import { LoginFormProps } from './login-form.props';
import { loginFormStyles as styles } from './login-form.styles';

/**
 * Login form
 *
 * @param loading Optional, true if logging in, disables button and show activity.
 * @param onPress Callback when login button is pressed.
 * @param style Optional, login form style override
 */
export const LoginForm: React.FC<LoginFormProps> = (props) => {
  const { onLogin, title } = props;

  return (
    <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
      <View style={styles.view.content}>
        <View style={styles.view.buttonContainer}>
          <Button title={title} onPress={onLogin} />
        </View>
      </View>
    </TouchableWithoutFeedback>
  );
};
