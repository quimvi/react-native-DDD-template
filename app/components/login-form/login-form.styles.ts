import { stylesheet } from '~/utils/styles';

export const ICON_LOGO_SIZE = 160;

export const loginFormStyles = stylesheet({
  view: {
    buttonContainer: {
      alignItems: 'flex-start',
      marginTop: 16,
    },
    content: {
      justifyContent: 'flex-end',
    },
  },
});
