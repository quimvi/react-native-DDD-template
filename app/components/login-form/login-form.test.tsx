import * as React from 'react';
import sinon from 'sinon';

import { render, fireEvent, RenderAPI, assert } from '~/test/utils';

import { LoginForm } from './login-form';

describe('LoginForm', function () {
  it('renders correctly', function () {
    const { toJSON } = render(<LoginForm title='send' />);
    expect(toJSON()).toMatchSnapshot();
  });

  describe('when the user press the login button', function () {
    it('the login function from props is called', async function () {
      const navigate = sinon.spy();

      const instance: RenderAPI = render(
        <LoginForm title='send' onLogin={navigate} />,
        {}
      );

      fireEvent(await instance.findByText('send'), 'press');
      assert.calledOnce(navigate);
    });
  });
});
