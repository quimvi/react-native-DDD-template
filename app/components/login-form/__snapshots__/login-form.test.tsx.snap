// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "justifyContent": "flex-end",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "marginTop": 16,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          send
        </Text>
      </View>
    </View>
  </View>
</View>
`;
